cql: Add support for IN clause on indexed column without filtering

Jager227 opened the pull request:
Fixes #21430

Jager227 commented:
@nyh I've removed xfail from the third test, as you asked. All tests have passed

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=21683

denesb commented:
I kicked off CI. If it passes, I'll queue this PR.

nyh commented:
@avikivity can you please review this patch? It is short, but changes the API of some of the CQL expression functions, for which you have the clearest vision and probably wouldn't like to see changes in the "wrong" direction.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13361/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/360/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4976/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1871/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12987/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13361/)
#### Failed Tests (12/35900):
- [cqlpy.cassandra_tests/validation/entities/secondary_index_test.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13361/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_secondary_index_test_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/secondary_index_test.debug.1)
- [cqlpy.cassandra_tests/validation/entities/secondary_index_test.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13361/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_secondary_index_test_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/secondary_index_test.debug.2)
- [cqlpy.cassandra_tests/validation/entities/secondary_index_test.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13361/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_secondary_index_test_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/secondary_index_test.debug.3)
- [cqlpy.cassandra_tests/validation/entities/secondary_index_test.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13361/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_secondary_index_test_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/secondary_index_test.dev.1)
- [cqlpy.cassandra_tests/validation/entities/secondary_index_test.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13361/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_secondary_index_test_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/secondary_index_test.dev.2)
- [cqlpy.cassandra_tests/validation/entities/secondary_index_test.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13361/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_secondary_index_test_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/secondary_index_test.dev.3)
- [cqlpy.cassandra_tests/validation/entities/secondary_index_test.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13361/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_secondary_index_test_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/secondary_index_test.release.1)
- [cqlpy.cassandra_tests/validation/entities/secondary_index_test.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13361/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_secondary_index_test_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/secondary_index_test.release.2)
- [cqlpy.cassandra_tests/validation/entities/secondary_index_test.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13361/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_secondary_index_test_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/secondary_index_test.release.3)
- [test_refuse_in_with_indexes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13361/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest_with_topology_changes___test_refuse_in_with_indexes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_refuse_in_with_indexes)
#### Build Details:
- Duration: 3 hr 57 min
- Builder: spider4.cloudius-systems.com

nyh commented:
Two tests failed in the CI:

The dtest cql_additional_tests.TestCQL.test_refuse_in_with_indexes, written by the Cassandra guys for https://issues.apache.org/jira/browse/CASSANDRA-4709, fails because it tries to enshrine the fact that IN on indexed column is not supported:
```
    @pytest.mark.single_node
    def test_refuse_in_with_indexes(self):
        """Test for the validation bug of #4709"""
    
        session = self.prepare()
        session.execute("create table t1 (pk varchar primary key, col1 varchar, col2 varchar);")
        session.execute("create index t1_c1 on t1(col1);")
        session.execute("create index t1_c2 on t1(col2);")
        session.execute("insert into t1  (pk, col1, col2) values ('pk1','foo1','bar1');")
        session.execute("insert into t1  (pk, col1, col2) values ('pk1a','foo1','bar1');")
        session.execute("insert into t1  (pk, col1, col2) values ('pk1b','foo1','bar1');")
        session.execute("insert into t1  (pk, col1, col2) values ('pk1c','foo1','bar1');")
        session.execute("insert into t1  (pk, col1, col2) values ('pk2','foo2','bar2');")
        session.execute("insert into t1  (pk, col1, col2) values ('pk3','foo3','bar3');")
>       assert_invalid(session, "select * from t1 where col2 in ('bar1', 'bar2');")
```

I think this dtest should be deleted - you can't do this in your patch and we'll need to do it in a separate patch, before this patch goes in.

Additionally, we have in test/cqlpy (in this repository) `cassandra_tests/validation/entities/secondary_index_test`::`testInvalidIndexSelect` which is the same test as the above dtest (the Cassandra guys converted it from dtest to Java, then years later I translated it to Python). I think this test should be commented out, adding a comment that we decided to support it in Scylla so it's no longer invalid (perhaps pointing to the relevant Scylla issue). This you can do yourself in this patch.

avikivity commented:
With no explanation in the patch changelog, I will not review this.

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=21683

Jager227 commented:
@avikivity I've added the changelog, sorry for the inconvenience. I've also commented out the `cassandra_tests/validation/entities/secondary_index_test::testInvalidIndexSelect` and added a comment as @nyh suggested.

nyh commented:
Sent a dtest patch to remove the duplicate test - https://github.com/scylladb/scylla-dtest/pull/5369.

avikivity commented:
> @avikivity I've added the changelog, sorry for the inconvenience. I've also commented out the `cassandra_tests/validation/entities/secondary_index_test::testInvalidIndexSelect` and added a comment as @nyh suggested.

The changelog should contain an example query that is enabled by this patch, to focus the reviewer.

avikivity commented:
There is a difficulty here related to mixed-version clusters. Suppose we have a pre-patch and a post-patch node in the same cluster. A query (with ALLOW FILTERING) starts on one of the nodes, gets paged, formats a paging_state thing and returns it to the client. The client now requests the next page from the other node. The other node generates a different query plan (with or without the index) and now misinterprets the paging state since (I think) the paging state format depends on the query plan.

Adding a cluster feature doesn't fix this, because we can have a paging state generated from before the cluster feature was agreed used after the cluster feature was agreed.

Probably the only way to fix this is to encode the query plan in paging state, and reuse the original query's query plan rather than re-planning each page. But that is a task much larger in scope than this patch.

Another alternative is to ignore the problem. @Jager227 does your workload have such queries already? They would be impacted during upgrade.

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -1081,22 +1081,24 @@ lw_shared_ptr<const service::pager::paging_state> indexed_table_select_statement
             return last_base_pk;
         } else {
-            return partition_key::from_single_value(*_view_schema, *indexed_column_value);
+            return partition_key::from_single_value(*_view_schema, *indexed_column_vector.back());
```
Let's add value_list_for that returns a vector, and change value_for() to use list_value_for() and verify it has exactly one element, avoiding the chance that we reference an empty vector or ignore some elements. It probably can't happen due to grammar constraints, but let's be defensive.

Jager227 replied:
I've designed value_for to return a vector of possible values for any operation on RHS, particularly IN operation.
For example, in "SELECT ck FROM tb WHERE x IN (2, 4)" query possible values would be 2 and 4. I don't see why we should check that value_for returns a single element. But, i agree that we could check if vector is empty to prevent errors.

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -1081,22 +1081,24 @@ lw_shared_ptr<const service::pager::paging_state> indexed_table_select_statement
 
     std::vector<managed_bytes_view> exploded_index_ck;
-    exploded_index_ck.reserve(_view_schema->clustering_key_size());
+    exploded_index_ck.reserve(_view_schema->clustering_key_size() * indexed_column_vector.size());
```
I don't understand this. A few lines back, you use indexed_column_vector.back(), indicating you expect one element. Here you treat it as a vector.

Jager227 replied:
I believe that we only need one value to create a partition key, but that's not the case with clustering keys. I've tested this setup and it works correct.

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -1286,14 +1288,15 @@ dht::partition_range_vector indexed_table_select_statement::get_partition_ranges
+            auto dk = dht::decorate_key(*_view_schema, pk);
+            auto range = dht::partition_range::make_singular(dk);
+            partition_ranges.emplace_back(range);
+        }
```
It's not very important, but might as well reserve space in partition_ranges.

avikivity commented:
Please add tests (or verify that such tests already exist) that show the new thing works with paging (for example use a page size of 2 to force paging).

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13391/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/377/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13007/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1888/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4995/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13391/)
#### Failed Tests (3/35905):
- [test_refuse_in_with_indexes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13391/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest___test_refuse_in_with_indexes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_refuse_in_with_indexes)
- [test_refuse_in_with_indexes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13391/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest_with_topology_changes___test_refuse_in_with_indexes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_refuse_in_with_indexes)
- [test_refuse_in_with_indexes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13391/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest_with_tablets___test_refuse_in_with_indexes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_refuse_in_with_indexes)
#### Build Details:
- Duration: 3 hr 50 min
- Builder: spider1.cloudius-systems.com
